plugin/src/main/kotlin/tanvd/grazi/language/LangDownloader.kt

```diff
@@ -0,0 +1,160 @@
+
+    private fun DependencyNode.traverse(action: (DependencyNode) -> Unit) {
+        action(this)
+        this.children.forEach(action)
+    }
+
+    fun downloadMissingLanguages(project: Project?) {
+        val state = GraziConfig.get()
+
+        if (state.hasMissedLanguages()) {
+            state.enabledLanguages.filter { it.jLanguage == null }.forEach {
+                with(LangDownloader) { it.downloadLanguage(project) }
+            }
+
+            if (state.nativeLanguage.jLanguage == null) {
+                with(LangDownloader) { state.nativeLanguage.downloadLanguage(project) }
+            }
+
+            ProjectManager.getInstance().openProjects.forEach {
+                DaemonCodeAnalyzer.getInstance(it).restart()
+            }
+        }
+    }
+
+    fun Lang.downloadLanguage(project: Project?): Boolean {
+        // check if language lib already loaded
+        if (GraziLibResolver.isLibExists("language-$shortCode-${msg("grazi.languagetool.version")}.jar")) {
+            return true
+        }
+
+        val jars: MutableList<Artifact> = ArrayList()
+        val isNotCancelled = CoreProgressManager.getInstance().runProcessWithProgressSynchronously({
+            val artifact = DefaultArtifact("org.languagetool", "language-$shortCode", "jar", msg("grazi.languagetool.version"))
+            val request = CollectRequest(artifact.createDependency(), listOf(MAVEN_CENTRAL_REPOSITORY))
+            try {
+                repository.collectDependencies(session, request).root.traverse { jars.add(it.artifact) }
+            } catch (e: Throwable) {
+                logger.trace("Download error", e)
+            }
+        }, msg("grazi.ui.settings.language.searching.title"), true, project)
+
+        // jars must have at least one jar with language
+        if (jars.isEmpty()) {
+            // FIXME very ugly
+            Messages.showWarningDialog(project, "Failed to download $displayName", "Download error")
+        }
+
+        if (isNotCancelled && jars.isNotEmpty()) {
+            val downloader = DownloadableFileService.getInstance()
+            val descriptions = jars.map { downloader.createFileDescription(it.url, it.name) }.toList()
+
+            val result = downloader.createDownloader(descriptions, "$displayName language")
+                    .downloadFilesWithProgress(GraziPlugin.installationFolder.absolutePath + "/lib", project, null)
+
+            // null if canceled or failed, zero result if nothing found
+            if (result != null && result.size > 0) {
+                with(UrlClassLoader::class.java.getDeclaredMethod("addURL", URL::class.java)) {
+                    isAccessible = true
+                    result.forEach {
+                        invoke(GraziPlugin.classLoader, Paths.get(it.presentableUrl).toUri().toURL())
+                    }
+                }
+
+                // register new Language at LanguageTool
+                with(Languages::class.java.getDeclaredField("dynLanguages")) {
```

Comment:
Hm... Move somewhere?... Something like `LanguageToolInstrumentation`... But I am not sure

Reply:
This is dirty hack, I don't really want to open it for freely use, also it's definitely not needed anywhere else. Probably better create an extension function like Lang.registerInLanguageTool?

Reply:
Yep, for example)

Reply:
Done